R/categorical_trees.R

```diff
@@ -0,0 +1,494 @@
+      !data.tree::GetAttribute(x, "exists")
+  )
+  if (!is.null(missing_nodes)) {
+    missing_nodes <- sort(unique(unname(missing_nodes)))
+  }
+  return(missing_nodes)
+}
+
+#' @rdname problematic_tree_nodes
+identify_missing_scale <- function(tree) {
+
+  # identify each node that is missing which causes scaling of itself or
+  # children to not be possible
+  missing_nodes <- tree$Get(
+    "name",
+    filterFun = function(x) !data.tree::GetAttribute(x, "exists") &
+      (!data.tree::GetAttribute(x, "scale_possible") |
+         !data.tree::GetAttribute(x, "scale_children_possible"))
+  )
+  if (!is.null(missing_nodes)) missing_nodes <- sort(unname(missing_nodes))
+  return(missing_nodes)
+}
+
+#' @rdname problematic_tree_nodes
+identify_present_agg <- function(tree) {
+
+  # identify each non leaf node that is present already when any of its
+  # children's nodes are present
+
+  # identify each nonleaf (and its subtree) where data exists
+  check_subtrees <- data.tree::Traverse(
+    tree, traversal = "post-order",
+    filterFun = function(x) {
+      data.tree::isNotLeaf(x) & data.tree::GetAttribute(x, "exists")
+    }
+  )
+
+  # check whether each subtree's parent has any children (or lower nodes) with
+  # data available
+  present_nodes <- lapply(check_subtrees, function(subtree) {
+    subtree_exists <- subtree$Get("exists")
+    lower_node_exists <- subtree_exists[names(subtree_exists) != subtree$name]
+    if (any(lower_node_exists)) {
+      return(subtree$name)
+    }
+    return(NULL)
+  })
+  present_nodes <- unlist(present_nodes)
+
+  return(present_nodes)
+}
+
+#' @title Create a list of subtrees that can be used during aggregation or
+#'   scaling
+#'
+#' @description
+#' Use [data.tree::Traverse()] to create a list of subtrees that are to be used
+#' during aggregation or scaling.
+#'
+#' For aggregation, traverses in post-order so that aggregation starts from
+#' the bottom of the tree and goes up. For scaling, traverses in pre-order so
+#' that scaling starts from the top of the tree and goes down.
+#'
+#' Only includes nodes that can be aggregated to or scaled.
+#'
+#' @inheritParams create_agg_tree
+#'
+#' @rdname create_subtrees
+create_agg_subtrees <- function(mapping, node_exists, col_type) {
+
+  tree <- create_agg_tree(mapping, node_exists, col_type)
+
+  # identify each nonleaf where aggregation is possible and its subtree
+  subtrees <- data.tree::Traverse(
+    tree, traversal = "post-order",
+    filterFun = function(x) {
+      data.tree::isNotLeaf(x) & data.tree::GetAttribute(x, "agg_possible")
+    }
+  )
+
+  # drop the last subtree since it is a mapping from the full interval to
+  # each aggregate which can't be used to aggregate. It is separately
+  # included as a subtree if needed
+  if (col_type == "interval" & tree$agg_possible) {
+    subtrees <- subtrees[-length(subtrees)]
+  }
+
+  return(subtrees)
+}
+
+#' @rdname create_subtrees
+create_scale_subtrees <- function(mapping,
```

Comment:
Do we test all of these helper functions or just exported functions?

Reply:
I was mainly testing the exported functions and some of the slightly more important and complicated helper functions.

We can add more tests down the line especially if we introduce bugs